update functors.md with minor tweaks (#2984)
ocaml/ocaml.org: data/tutorials/language/1ms_01_functors.md

[Dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is a way to parametrise over a dependency.

> Here is a refactoring of the module `IterPrint` to use dependency injection:

Reviewer: Why the repetition here?

end
```

The module `IterPrint` is refactored into a functor that takes a module providing the function `iter` as a parameter. The `with type 'a t := 'a Dep.t` is called a "destructive type substitution", and it is a constraint that means the type `t` from the parameter `Dep` replaces the type `t` in the result module. This allows `f`'s type to use `Dep`'s `t` type. With this refactoring, `IterPrint` only has one dependency. At its compilation time, no implementation of function `iter` is available yet.

> **Note**: While an OCaml interface file (`.mli`) is a module, an interface file is not allowed to be a functor. Functors must be embedded inside modules. Therefore, it is customary to call them `Make`. <!-- I find this confusing; is "Make" a synononym for OCaml functors? -Jakub -->

Reviewer 2: ```suggestion
**Note**: An OCaml file (`.ml`) defines a module, not a functor. Functors must be embedded inside modules. It is customary to call the functor `Make`. This allows writing `module IntSet = Set.Make(Int)`.
```

You're right; the initial text was confusing and wrong. Thanks for pointing this. Do you think the code `module IntSet = Set.Make(Int)` needs to be explained?

Reviewer: Probably not a bad idea.

end
```

In the example above, `t` from `with type` takes precedence over the local `t`, which only has a local scope. 

> **Warning**: Don't shadow names too often because it makes the code harder to understand.

Reviewer 2: Good point!
